### ISSUE TITLE
Consider using Joi for schema validation

### ISSUE THREAD

**Issue by marshall007** [neumino/thinky#277]
I believe the existing schema API is already based heavily on [`Joi`](https://github.com/hapijs/joi). I think it would make a lot of sense to just move to it completely.

1. Reduce complexity of `thinky`.
1. `Joi` can be browserify'd, so schemas could be shared between client and server (this would be huge!).
1. Many applications require validating things other than database models. It would be nice to use a single library for both cases.

**Comment by MarkHerhold** [comment]
+1
I love Joi and it would be wonderful if Joi used it. I remember seeing thinky's validation and thinking how crazy it was that thinky is implementing its own type checking and isn't just using Joi. :laughing:

**Comment by Tlvenn** [comment]
+1 for this as well.

**Comment by neumino** [comment]
Yes, the current syntax is based on Joi. I don't remember who showed it to me, but I found the syntax way nicer that what we used to have.

I don't remember exactly all the reasons, but the main one I think was that ReQL values are not checked on `save` but on retrieval. Typicall `r.now()` will pass the validation for `save`.

**Comment by neumino** [comment]
That being said, I'm not an expert on Joi and may have missed an option to skip validation in special cases (or they may have added it since last time?).

**Comment by robertjpayne** [comment]
@neumino does Joi need to skip all validation or just validation on certain fields?

**Comment by neumino** [comment]
It needs to skip validation sometimes (for ReQL terms like `r.now()`), or have a special handling for binaries/dates when they are in the ReQL raw format.

**Comment by grantcarthew** [comment]
+1

**Comment by marshall007** [comment]
I think the solution to validating on retrieval would be for `thinky` to inspect the Joi schema you pass in to create the Model and replace (for example) instances of `Joi.date()` (i.e. `schema._type === 'date'`) with:

```js
Joi.alternatives().try(
  Joi.date(),
  Joi.object().keys({
    '$reql_type$': 'TIME'
  }).and('epoch_time', 'timezone')
)
```

I'm still unsure how we'd get Joi to not complain about validating `r.now()` before save, though.

**Comment by marshall007** [comment]
On second thought, it might be as simple as adding another alternative `Joi.object().type(r._Term)`?

**Comment by sjmueller** [comment]
+1

**Comment by cur3n4** [comment]
+1

**Comment by no-more** [comment]
That would really be great!

**Comment by sandeepjain** [comment]
Any chance if this is implemented? Any fork implementing this?

**Comment by at0g** [comment]
👍  Would love to see this

**Comment by schettino** [comment]
+1

**Comment by MarkHerhold** [comment]
@schettino @at0g please stop spamming this thread. Use the reaction buttons above

**Comment by no-more** [comment]
Does anyone know if there a tool to export thinky schema to joi schema ? I understand thinky have some requirements that joi don't have due to the database (for date), but it would be great to have a tool, even if it does not handle every options, that could export to joi which can then be used for routes validation.

**Comment by cedbale** [comment]
+1
